File: ios/Packages/FeatureServices/ConnectionStatusService/ConnectionStatusObserver.swift
// Copyright (c). Gem Wallet. All rights reserved.

import class Gemstone.GemConnectionService
import GemstonePrimitives
import Observation
import Primitives

@Observable
@MainActor
public final class ConnectionStatusObserver {
    private let connectionService: GemConnectionService
    private let monitors: [any ConnectionComponentMonitoring]

    public private(set) var isHealthyByComponent: [ConnectionComponent: Bool] = [:] {
        didSet {
            if status != status(for: oldValue) {
                debugLog("Connection status changed: \(status)")
            }
        }
    }

    @ObservationIgnored private var tasks: [Task<Void, Never>] = []

    public nonisolated init(
        connectionService: GemConnectionService,
        monitors: [any ConnectionComponentMonitoring],
    ) {
        self.connectionService = connectionService
        self.monitors = monitors
    }

    public func start() {
        guard tasks.isEmpty else { return }
        tasks = monitors.map { monitor in
            Task { [weak self] in
                for await isHealthy in monitor.healthStream() {
                    self?.update(component: monitor.component, isHealthy: isHealthy)
                }
            }
        }
    }

    public func stop() {
        tasks.forEach { $0.cancel() }
        tasks = []
    }

    public var status: ConnectionStatus {
        status(for: isHealthyByComponent)
    }

    func update(component: ConnectionComponent, isHealthy: Bool) {
        if connectionService.resetsComponentHealth(component: component.map(), isHealthy: isHealthy, wasHealthy: isHealthyByComponent[component]) {
            isHealthyByComponent = [:]
        }
        isHealthyByComponent[component] = isHealthy
    }

    private func status(for components: [ConnectionComponent: Bool]) -> ConnectionStatus {
        Array(components.filter { !$0.value }.keys).connectionStatus
    }
}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/connection/ConnectionStatusObserver.kt
package com.gemwallet.android.data.services.gemstone.connection

import com.gemwallet.android.ext.toGem
import com.gemwallet.android.serializer.toJson
import com.wallet.core.primitives.ConnectionComponent
import com.wallet.core.primitives.ConnectionStatus
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import uniffi.gemstone.GemConnectionService
import uniffi.gemstone.GemConnectionServiceInterface
import uniffi.gemstone.GemRefreshKind

class ConnectionStatusObserver(private val monitors: List<ConnectionComponentMonitor>, private val connectionService: GemConnectionServiceInterface, private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)) {
    private val state = MutableStateFlow<Map<ConnectionComponent, Boolean>>(emptyMap())

    val isHealthyByComponent: StateFlow<Map<ConnectionComponent, Boolean>> = state.asStateFlow()

    val status: StateFlow<ConnectionStatus> = state
        .map { it.connectionStatus }
        .distinctUntilChanged()
        .stateIn(scope, SharingStarted.Eagerly, ConnectionStatus.Online)

    private var jobs: List<Job> = emptyList()

    fun refreshIntervalMillis(kind: GemRefreshKind): Flow<Long> = status.map { connectionService.refreshInterval(kind, it.toGem()).toMillis() }

    fun start() {
        if (jobs.isNotEmpty()) return
        jobs = monitors.map { monitor ->
            scope.launch {
                monitor.healthFlow().collect { isHealthy ->
                    update(monitor.component, isHealthy)
                }
            }
        }
    }

    fun stop() {
        jobs.forEach { it.cancel() }
        jobs = emptyList()
    }

    internal fun update(component: ConnectionComponent, isHealthy: Boolean) {
        state.update { current ->
            val base = if (connectionService.resetsComponentHealth(component.toGem(), isHealthy, current[component])) emptyMap() else current
            base + (component to isHealthy)
        }
    }
}
